### PR TITLE
Added support for floats

### PR DESCRIPTION
**Pull request by torch2424**
Hello! 

This PR adds support for parsing and handling floats. :smile: :tada: 

The old `Number` type becomes an alias for `Float`, similar to the AssemblyScript SDK: https://www.assemblyscript.org/stdlib/number.html#number

All tests are passing, and I added some tests to differentiate Numbers, Integers, and Floats :smile: :tada: 

My plan after this PR, is to tidy up the README. Let me know what you think, and if you need more info! :smile: :tada: 

Screenshot from 2021-01-20 16-47-22

### CODE REVIEW

File: assembly/decoder.ts [hunk 1]
```diff
@@ -301,22 +313,41 @@ export class JSONDecoder<JSONHandlerT extends JSONHandler> {
...
+        if (decimalPlace == 1) {
+          number *= 10;
+          number += value;
+        } else {
+          value *= decimalPlace;
+          number += value;
+          decimalPlace = decimalPlace / 10;
+        }
+        digits++;
+      }
```
Comment (MaxGraey) [inline assembly/decoder.ts:343]: I recommend read whole float and use `parseFloat` method due to this custom parser will be very imprecise

File: assembly/decoder.ts [hunk 2]
```diff
@@ -301,22 +313,41 @@ export class JSONDecoder<JSONHandlerT extends JSONHandler> {
...
+        let byte = this.readChar();
+        let value: f64 = byte - CHAR_0;
+        if (decimalPlace == 1) {
+          number *= 10;
+          number += value;
+        } else {
+          value *= decimalPlace;
+          number += value;
+          decimalPlace = decimalPlace / 10;
+        }
```
Comment (dcodeIO) [inline assembly/decoder.ts:341]: May lose precision here I think due to floating point peculiarities. For better precision, one could validate that the parsed string is a correctly formatted float literal here, and then pass the entire literal through `parseFloat` which does the typical scaling and whatnot to retain accuracy.

Comment (dcodeIO) [reply assembly/decoder.ts:341]: Seems Max and I had the same idea for the comment, closing mine :)

**Comment by torch2424** [comment]
@MaxGraey Made requested changes :smile: 

cc @willemneal I'd also be stoked to see your review :smile: :tada:

File: assembly/decoder.ts [hunk 3]
```diff
@@ -315,38 +315,34 @@ export class JSONDecoder<JSONHandlerT extends JSONHandler> {
...
-          number += value;
-          decimalPlace = decimalPlace / 10;
         }
         digits++;
       }
     }
     if (digits > 0) {
-      if (decimalPlace == 1) {
-        this.handler.setInteger(this.state.lastKey, <i64>(number * sign));
+      if (numberAsString.includes(".")) {
```
Comment (MaxGraey) [inline assembly/decoder.ts:342]: I guess make sense also check `e` and `E`

Comment (torch2424) [reply assembly/decoder.ts:342]: @MaxGraey just to confirm, you are reffering to scientific notation, like: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toExponential ? :smile:

Comment (MaxGraey) [reply assembly/decoder.ts:342]:
```ts
let isFloat = false
...
let charCode = this.readChar();
if (charCode == Chars.Dot || charCode == Chars.e || charCode == Chars.E) {
   isFloat = true;
}
numberAsString += String.fromCharCode(charCode);
```

Comment (MaxGraey) [reply assembly/decoder.ts:342]: > just to confirm, you are reffering to scientific notation

Yes

**Comment by torch2424** [comment]
@MaxGraey Made the requested changes! :smile: :tada:

**Comment by MaxGraey** [comment]
I see a lot of edge cases, but for basic supports should be enought

File: assembly/__tests__/roundtrip.spec.ts [hunk 4]
```diff
@@ -32,10 +32,19 @@ describe("Round trip", () => {
...
+  });
+
   it("should handle int32Sign", () => {
     roundtripTest('{"int":-4660}');
   });
 
+  it("should handle float32Sign", () => {
+    roundtripTest('{"float":-24.24}');
+  });
+
```
Comment (MaxGraey) [inline assembly/__tests__/roundtrip.spec.ts:47]: Could you also add tests for scientific float point numbers?

**Comment by torch2424** [comment]
@MaxGraey Made your requested changes, I also recognized that we also should support `e+` and `E+`, as AS supports parsing this as well, and it's part of the JS behavior: https://developer.mozilla.org/en-US/docs/Web/CSS/number

Screenshot from 2021-01-25 16-32-48

File: assembly/decoder.ts [hunk 5]
```diff
@@ -342,7 +345,7 @@ export class JSONDecoder<JSONHandlerT extends JSONHandler> {
       let charCode = this.readChar();
       numberAsString += String.fromCharCode(charCode);
 
-      if (charCode == CHAR_E || charCode == CHAR_E_LOWER || charCode == CHAR_PERIOD || charCode == CHAR_PLUS) {
+      if (charCode == CHAR_E || charCode == CHAR_E_LOWER || charCode == CHAR_PERIOD || charCode == CHAR_PLUS || charCode == CHAR_MINUS) {
```
Comment (MaxGraey) [inline assembly/decoder.ts:348]: Hmm, but `+123` and `-123` could be integer as well. I guess better detect `e+` / `E+` and `e-` / `E-`

**Comment by torch2424** [comment]
@MaxGraey Made requested changes, and with tests! :smile: :tada:

File: assembly/decoder.ts [hunk 6]
```diff
@@ -321,6 +338,25 @@ export class JSONDecoder<JSONHandlerT extends JSONHandler> {
   }
 
   private parseNumber(): bool {
+
+    // Check for our special strings
+    // Peek out the characters of our longest number string
+    let numberPeekString = this.peekString(NEGATIVE_INFINITY_STR.length);
+
+    if (numberPeekString.indexOf(NaN_STR) === 0) {
```
Comment (MaxGraey) [inline assembly/decoder.ts:346]: `numberPeekString.startsWith(NaN_STR)`

**Comment by torch2424** [comment]
@MaxGraey More fixes haha! :smile: 

Screenshot from 2021-01-25 17-20-54

**Comment by MaxGraey** [comment]
I apologise! It seems `NaN` and `Infinity` is not valid numeric literals in JSON =) That's pretty strange, but NaN/Infinity will stringify as `null` and actually invalid tokens during parsing

**Comment by MaxGraey** [comment]
So `JSON.parse("-0")` is fine and supported. But support for NaN and Infinity it seems unnecessary

**Comment by torch2424** [comment]
Ah, Alright I'll remove support for NaN and Infinity :joy:

**Comment by torch2424** [comment]
@MaxGraey Made requested changes, this should be good to go! :smile:

**Comment by MaxGraey** [comment]
LGTM. Just one question. It seems Number and Float classes the same could you leave only one of them?

**Comment by torch2424** [comment]
@MaxGraey I don't think I can remove them, since `JSON.Number` was exposed from the API, it'd be a breaking change if I were to remove the name entirely :disappointed: 

So to avoid that, I think it'd be best practice just to keep it in, as an alias to `Float`, like we do in the AS Stdlib :smile:

**Comment by MaxGraey** [comment]
ok

**Comment by torch2424** [comment]
Awesome! (P.S thank you for the very thorough review! I super appreciate it @MaxGraey ! :smile: :pray: :tada: )

**Review by willemneal — APPROVED** [review]

**Pull request merged**